validator/server_arb/machine_cache.go

```diff
@@ -56,156 +58,225 @@ func NewMachineCache(ctx context.Context, initialMachineGetter func(context.Cont
+	if newInterval == 0 {
+		newInterval = 1
+	}
 	if start == c.firstMachineStep && newInterval == c.machineStepInterval {
-		return c
+		return nil
 	}
-	newCache := &MachineCache{
-		Promise: containers.NewPromise[struct{}](),
-		config:  c.config,
+	closestIndex, closest := c.getClosestMachine(start)
+	closestStep := closest.GetStepCount()
+	if closestStep > start {
+		return fmt.Errorf("initial machine step too large %d > %d", closestStep, start)
 	}
-	go func() {
-		_, err := c.Await(ctx)
-		if err != nil {
-			newCache.ProduceError(err)
-			return
+	for i, mach := range c.machines {
+		if i != closestIndex {
+			mach.Destroy()
 		}
-		newCache.zeroStepMachine = c.zeroStepMachine
-		newCache.finalMachine = c.finalMachine
-		closest, err := c.getClosestMachine(start)
+	}
+	var initial MachineInterface
+	if closestStep < start {
+		initial = closest.CloneMachineInterface()
+		err := initial.Step(ctx, start-closestStep)
 		if err != nil {
-			newCache.ProduceError(err)
-			return
-		}
-		closestStep := closest.GetStepCount()
-		var initial MachineInterface
-		if closestStep > start {
-			newCache.ProduceError(fmt.Errorf("initial machine step too large %d > %d", closestStep, start))
-			return
-		}
-		if closestStep < start {
-			initial = closest.CloneMachineInterface()
-			err := initial.Step(ctx, start-closestStep)
-			if err != nil {
-				newCache.ProduceError(err)
-				return
-			}
-			initial.Freeze()
-		} else {
-			initial = closest
+			return err
 		}
-		newCache.machines = []MachineInterface{initial}
-		newCache.firstMachineStep = start
-		newCache.machineStepInterval = newInterval
-		err = newCache.populateInitialCache(ctx, newInterval*uint64(c.config.CachedChallengeMachines))
-		if err != nil {
-			newCache.ProduceError(err)
-		} else {
-			newCache.Produce(struct{}{})
+		if closest != c.zeroStepMachine && closest != c.finalMachine {
+			closest.Destroy()
 		}
-	}()
-	return newCache
+		initial.Freeze()
+	} else {
+		initial = closest
+	}
+	c.machines = []MachineInterface{initial}
+	c.firstMachineStep = start
+	c.machineStepInterval = newInterval
+	return c.populateCache(ctx)
+}
+
+func (c *MachineCache) SetRange(ctx context.Context, start uint64, end uint64) error {
+	err := c.lockBuild(ctx)
+	if err != nil {
+		return err
+	}
+	err = c.setRangeLocked(ctx, start, end)
+	c.unlockBuild(err)
+	return err
 }
 
-func (c *MachineCache) populateInitialCache(ctx context.Context, target_step uint64) error {
+func (c *MachineCache) populateCache(ctx context.Context) error {
+	var nextMachine MachineInterface
+	if len(c.machines) == 0 {
+		nextMachine = c.zeroStepMachine
+		c.firstMachineStep = c.machineStepInterval
+	} else {
+		nextMachine = c.machines[len(c.machines)-1]
+	}
 	for {
-		nextMachine := c.machines[len(c.machines)-1].CloneMachineInterface()
 		if !nextMachine.IsRunning() {
 			break
 		}
-		if nextMachine.GetStepCount() >= target_step {
+		if nextMachine.GetStepCount() >= c.finalMachineStep {
```

Comment:
I think we should instead check if `nextMachine.GetStepCount() + c.machineStepInterval >= c.finalMachineStep`. I don't think we want another final step machine in the machines list.